3.12/tier2.md

```diff
@@ -0,0 +1,466 @@
+
+```
+  f:
+    RESUME                   0
+    
+    LOAD_FAST__LOAD_CONST    0 (y)
+    LOAD_CONST               1 (2)
+    BINARY_OP_MULTIPLY_INT   5 (*)
+    RETURN_VALUE
+
+  sumf:
+    RESUME_QUICK             0
+    
+    LOAD_CONST               1 (0)
+    STORE_FAST__LOAD_FAST     1 (t)
+    
+    LOAD_FAST                0 (x)
+    GET_ITER
+  loop:
+    FOR_ITER_RANGE          18 (to 50)
+    STORE_FAST__LOAD_FAST     2 (y)
+    
+    LOAD_FAST                1 (t)
+    LOAD_GLOBAL_MODULE       1 (NULL + f)
+    LOAD_FAST                2 (y)
+    CALL_PY_EXACT_ARGS       1
+    BINARY_OP_ADD_INT       13 (+=)
+    STORE_FAST               1 (t)
+    JUMP_BACKWARD_QUICK     loop
+    
+    LOAD_CONST               0 (None)
+    RETURN_VALUE
+```
+
+Assuming that the counter triggers at `JUMP_BACKWARD_QUICK`
+then the trace starts at `FOR_ITER_RANGE`,
+we can then project the trace to `CALL_PY_EXACT_ARGS`
+as there are no branches, and the specialized instructions
+have good hit rates (100% in this case).
+We can even project the trace into `f` as `CALL_PY_EXACT_ARGS`
+has 100% hit rates. Resulting in the following trace:
+
+```
+  start:
+    SAVE_LASTI
+    FOR_ITER_RANGE           exit_trace
+    SAVE_LASTI
+    STORE_FAST               2 (y)
+    SAVE_LASTI
+    LOAD_FAST                1 (t)
+    SAVE_LASTI
+    LOAD_GLOBAL_MODULE       1 (NULL + f)
+    SAVE_LASTI
+    LOAD_FAST                2 (y)
+    SAVE_LASTI
+    CALL_PY_EXACT_ARGS       1
+    SAVE_LASTI
+    LOAD_FAST                0 (y)
+    SAVE_LASTI
+    LOAD_CONST               1 (2)
+    SAVE_LASTI
+    BINARY_OP_MULTIPLY_INT   5 (*)
+    SAVE_LASTI
+    RETURN_VALUE
+    SAVE_LASTI
+    BINARY_OP_ADD_INT       13 (+=)
+    SAVE_LASTI
+    STORE_FAST               1 (t)
+    SAVE_LASTI
+    JUMP_BACKWARD_QUICK      start
+  exit_trace:
+    EXIT_TO_INTERPRETER
+```
+
+Note the addition of `SAVE_LASTI` in between each instruction.
+This is so that when an instruction deopts, it will return to the correct location in the original bytecode.
+
+## Optimizing traces
+
+Optimizing traces is a well research area. Rather than regurgitate that
+research we run through how the above trace would be optimized,
+as an example.
+
+First of all we lower the specialized instructions into their
+two parts; guard then action:
+
+```
+  start:
+    SAVE_LASTI
+    CHECK_IS_RANGE           deopt1
+    FOR_ITER_RANGE_UNCHECKED exit_trace
+    SAVE_LASTI
+    STORE_FAST               2 (y)
+    SAVE_LASTI
+    LOAD_FAST                1 (t)
+    SAVE_LASTI
+    CHECK_GLOBAL_KEYS_VERSION deopt2
+    LOAD_GLOBAL_MODULE_UNCHECKED 1 (NULL + f)
+    SAVE_LASTI
+    LOAD_FAST                2 (y)
+    SAVE_LASTI
+    CHECK_PY_EXACT_ARGS      1 (f)
+    ENTER_PY_FUNCTION        1 (f)
+    SAVE_LASTI
+    LOAD_FAST                0 (y)
+    SAVE_LASTI
+    LOAD_CONST               1 (2)
+    SAVE_LASTI
+    CHECK_INTS_2             deopt3
+    BINARY_OP_MULTIPLY_INT_UNCHECKED (*)
```

Comment:
There are a handful these asterisks. What do they mean?

Reply:
`a * b` It's a multiply instruction